external-crates/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -97,11 +99,164 @@ impl<'a, 'b> TypeView for TypeWithLoader<'a, 'b> {
     }
 }
 
+pub trait InterpreterInterface {
+    fn get_stack_len(&self) -> usize;
+    fn get_internal_state(&self) -> ExecutionState;
+    fn set_location(&self, err: PartialVMError) -> VMError;
+    fn maybe_core_dump(&self, err: VMError, frame: &Frame) -> VMError;
+}
+
+impl InterpreterInterface for Interpreter {
+    fn get_stack_len(&self) -> usize {
+        self.get_stack_len()
+    }
+    fn get_internal_state(&self) -> ExecutionState {
+        self.get_internal_state()
+    }
+
+    fn set_location(&self, err: PartialVMError) -> VMError {
+        self.set_location(err)
+    }
+
+    fn maybe_core_dump(&self, err: VMError, frame: &Frame) -> VMError {
+        self.maybe_core_dump(err, frame)
+    }
+}
+
 impl Interpreter {
     /// Limits imposed at runtime
     pub fn runtime_limits_config(&self) -> &VMRuntimeLimitsConfig {
         &self.runtime_limits_config
     }
+
+    pub fn handle_plugin_error(result: VMResult<()>, is_critical: bool) -> VMResult<()> {
+        match result {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                if is_critical {
+                    return Err(err);
+                } else {
+                    error!("Plugin error: {:?}", err);
+                }
+                Ok(())
+            }
+        }
+    }
+
+    pub fn pre_entrypoint(
+        plugins: &mut Vec<Box<dyn Plugin>>,
+        gas_meter: &mut impl GasMeter,
+        function: &Arc<Function>,
+        ty_args: &[Type],
+        data_store: &mut impl DataStore,
+        loader: &Loader,
+    ) -> VMResult<()> {
+        profile_open_frame!(gas_meter, function.pretty_string());
+        let resolver = function.get_resolver(data_store.link_context(), loader);
+
+        for plugin in plugins.iter_mut() {
+            let result = plugin.pre_entrypoint(function, ty_args, &resolver);
+            Self::handle_plugin_error(result, plugin.is_critical())?;
+        }
+
+        Ok(())
+    }
+
+    pub fn post_entrypoint(plugins: &mut Vec<Box<dyn Plugin>>) -> VMResult<()> {
+        for plugin in plugins.iter_mut() {
+            let result = plugin.post_entrypoint();
+            Self::handle_plugin_error(result, plugin.is_critical())?;
+        }
+
+        Ok(())
+    }
+
+    pub fn pre_fn(
+        interpreter: &mut Interpreter,
+        plugins: &mut Vec<Box<dyn Plugin>>,
+        current_frame: &mut Frame,
+        data_store: &mut impl DataStore,
+        loader: &Loader,
+        gas_meter: &mut impl GasMeter,
+        function: &Arc<Function>,
+        ty_args: &[Type],
+    ) -> VMResult<()> {
+        #[cfg(debug_assertions)]
+        profile_open_frame!(gas_meter, function.pretty_string());
```

Comment:
should this be implemented as a `Plugin`?

Reply:
Yes. I'll pick this up in the gas refactor